addons/sale/models/sale.py

```diff
@@ -485,8 +485,14 @@ def update_prices(self):
             price_unit = self.env['account.tax']._fix_tax_included_price_company(
                 line._get_display_price(product), line.product_id.taxes_id, line.tax_id, line.company_id)
             if self.pricelist_id.discount_policy == 'without_discount' and price_unit:
-                price_discount_unrounded = self.pricelist_id.get_product_price(product, line.product_uom_qty, self.partner_id, self.date_order, line.product_uom.id)
-                discount = max(0, (price_unit - price_discount_unrounded) * 100 / price_unit)
+                price_discount_unrounded, rule_id = line.order_id.pricelist_id.with_context(product.env.context).get_product_price_rule(
+                    line.product_id,
+                    line.product_uom_qty or 1.0,
+                    line.order_id.partner_id
+                )
+                new_list_price, __ = line.with_context(product.env.context)._get_real_price_currency(product, rule_id,
+                                        line.product_uom_qty, line.product_uom, line.order_id.pricelist_id.id)
+                discount = max(0, (new_list_price - price_discount_unrounded) * 100 / new_list_price)
             else:
                 discount = 0
             lines_to_update.append((1, line.id, {'price_unit': price_unit, 'discount': discount}))
```

Comment:
Well, I'm sick of this duplicated code (and the unending bugfixes on it), can't we simply call the onchanges here ?
```python
for line in self.order_line.filtered(...):
    line.product_uom_change()
    line.discount = 0  # Force 0 as discount for the cases when _onchange_discount directly returns
    line._onchange_discount()
self.show_update_pricelist = False
...
```

Also, please add a test to cover your case ;)